### PR TITLE
Add QgsCPLHTTPFetchOverrider class and use it in OGR provider 

### PR DESCRIPTION
**Pull request by rouault**
This utility class redirects GDAL's CPL HTTP calls through QgsBlockingNetworkRequest,
for GDAL >= 3.2 (see OSGeo/gdal#2991)

And also QgsBlockingNetworkRequest: add head(), put() and deleteResource() methods

### CODE REVIEW

**Comment by nyalldawson** [comment]
This is great! 

One improvement I'd like to see here relates to the QgsSetRequestInitiatorClass macro which automatically sets the source location where a request is created, and it's used for transparency of network requests (it's visible in the network logger panel). I think it should be possible to create a similar macro which creates the QgsCPLHTTPFetchOverrider object while also using QgsSetRequestInitiatorClass to tag the source location in the request.

File: src/core/providers/ogr/qgsogrprovider.cpp [hunk 1]
```diff
@@ -512,6 +513,17 @@ QgsOgrProvider::QgsOgrProvider( QString const &uri, const ProviderOptions &optio
                           mSubsetString,
                           mOgrGeometryTypeFilter );
 
+  if ( mFilePath.contains( QLatin1String( "authcfg" ) ) )
+  {
+    QRegularExpression authcfgRe( " authcfg='([^']+)'" );
```
Comment (nyalldawson) [inline src/core/providers/ogr/qgsogrprovider.cpp:518]: Is the space at the start deliberate?

File: src/core/qgsblockingnetworkrequest.h [hunk 2]
```diff
@@ -106,6 +106,66 @@ class CORE_EXPORT QgsBlockingNetworkRequest : public QObject
...
+     *
+     * The optional \a feedback argument can be used to abort ongoing requests.
+     *
+     * The method will return NoError if the get operation was successful. The contents of the reply can be retrieved
+     * by calling reply().
+     *
+     * If an error was encountered then a specific ErrorCode will be returned, and a detailed error message
+     * can be retrieved by calling errorMessage().
+     *
+     * @since 3.18
```
Comment (nyalldawson) [inline src/core/qgsblockingnetworkrequest.h:127]: @since -> \since

File: src/core/providers/ogr/qgsogrprovider.cpp (hunk 1 above)
Comment (rouault) [reply src/core/providers/ogr/qgsogrprovider.cpp:518]: yes. Taken from another occurence in that file

File: src/core/qgsblockingnetworkrequest.h (hunk 2 above)
Comment (rouault) [reply src/core/qgsblockingnetworkrequest.h]: done

**Comment by rouault** [comment]
>  I think it should be possible to create a similar macro which creates the QgsCPLHTTPFetchOverrider object while also using

done

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
@rouault can you fix the conflicts here and then merge?

**Pull request merged**